### DISCUSSION TITLE
How did you add linked headers to section header?

### DISCUSSION THREAD

**Discussion by serhii-bilyk-intellias** [mdx-js/mdx#2350] [Q&A] [answered]
![guides](https://github.com/mdx-js/mdx/assets/144336942/4069d5b0-f14d-48f0-8f90-0aee88491ce9)

When I am using this syntax:
```md 
# Any section
```
Any section header is created without ID and link.
It is 
```html
<h1>Any section </h1>
```
And I want smth like that:
```html
<h1 id="batman"><a href="#batman">Any section</a></h1>
```

**Comment by wooorm** [comment, marked as answer]
https://github.com/rehypejs/rehype-autolink-headings

**Reply by serhii-bilyk-intellias** [reply]
Thank you @wooorm , I am very new to this ecosystem, I dont even understand how to make it work with webpack.
Getting this:
```js
[webpack-cli] Failed to load webpack.config.js' config
[webpack-cli] Error [ERR_REQUIRE_ESM]: require() of ES Module \node_modules\rehype-autolink-headings\index.js from webpack.config.js not supported.
```

**Reply by wooorm** [reply]
That‘s an ESM error. All packages are ESM. Click the “ESM only” link in the readme for more info on ESM: https://github.com/rehypejs/rehype-autolink-headings#install.

**Reply by serhii-bilyk-intellias** [reply]
@wooorm Thank you for your answer. I was able to make webpack config compile but this config does not make heading linked:

```javascript

import rehype from "rehype-autolink-headings";
module.export =...
      {
        test: /\.mdx?$/,
        use: [
          {
            loader: "@mdx-js/loader",
            /** @type {import('@mdx-js/loader').Options} */
            options: {
              rehypePlugins: [rehype],
            },
          },
        ],
      },
```
I don't understand how to make rehype plugin work with react + webpack + mdx. Is there any example how to configure it ? I have found a lot of example with `.use()` but whis is not for webpack

**Reply by wooorm** [reply]
I mentioned [`rehypejs/rehype-autolink-headings`](https://github.com/rehypejs/rehype-autolink-headings). That’s a plugin. `rehype` itself is not a plugin.

The field is called `rehypePlugins`. You put rehype plugins in there. `rehypeAutolinkHeadings` is a plugin. You import that, and put it in `rehypePlugins`

**Reply by wooorm** [reply]
oh wait, you named `rehype-autolink-headings` as `rehype`. That threw me off. I recommend using the name of the plugin as the identifier, instead of something else:

```
import rehypeAutolinkHeading from "rehype-autolink-headings";

...

              rehypePlugins: [rehypeAutolinkHeading],
```

That should work. If not, explain in more detail.

**Reply by serhii-bilyk-intellias** [reply]
@wooorm maybe I should include other loaders or plugins ?
```javascript
import rehypeAutolinkHeading from "rehype-autolink-headings";

      {
        test: /\.mdx?$/,
        use: [
          {
            loader: "@mdx-js/loader",
            /** @type {import('@mdx-js/loader').Options} */
            options: {
              rehypePlugins: [rehypeAutolinkHeading],
            },
          },
        ],
      },
```
This config just returns `<h1>Text</h1>`

**Reply by wooorm** [reply]
Please spend some time expanding on your question. See https://mdxjs.com/community/support/#asking-quality-questions for tips. Particularly, input MDX is useful.

**Reply by serhii-bilyk-intellias** [reply]
@wooorm I want to make heading linked. This is my mdx
```
## Table of Contents
```
I expect it to generate 
```html
<h1><a href="#">text</a></h1>
```
Instead I am getting 
```html
<h1>text</h1>
```
This is a config I am using in webpack 
```javascript
      {
        test: /\.mdx?$/,
        use: [
          {
            loader: "@mdx-js/loader",
            /** @type {import('@mdx-js/loader').Options} */
            options: {
              rehypePlugins: [rehypeAutolinkHeading],
            },
          },
        ],
      },
```

**Reply by serhii-bilyk-intellias** [reply]
Were able to make it work. Few notes: `rehype` plugin does not work if webpack config is a typescript file. In order to make it work, I had to change webpack config extension to `webpack.config.mjs`, also added `"type":"module"` to package json. This is my config:
```javascript
      {
        test: /\.mdx?$/,
        use: [
          { loader: "babel-loader", options: {} },

          {
            loader: "@mdx-js/loader",
            /** @type {import('@mdx-js/loader').Options} */
            options: {
              rehypePlugins: [
                rehypeSlug,
                [rehypeAutolinkHeading, { behavior: "wrap" }],
              ],
            },
          },
        ],
      },
```

**Reply by ChristianMurphy** [reply]
> Were able to make it work. Few notes: rehype plugin does not work if webpack config is a typescript file. 

It will if you have TypeScript configured to use Node 16 module loading https://github.com/mdx-js/mdx/blob/07e2b36edf6bbbf7b1c94134a8ab06a34967010d/tsconfig.json#L9
And _potentially_ if you use an `.mts` file https://www.typescriptlang.org/docs/handbook/esm-node.html

---

> `"type":"module"`

I'd recommend keeping this.

> extension to `webpack.config.mjs`

both `.mjs` and `.mts` should be unnecessary if you add `"type":"module"` to your _package.json_ and ` "module": "node16"` to you _tsconfig.json_
